### PR TITLE
Allow duplicate weighted backend keys

### PR DESCRIPTION
**Pull request by sfleen**
Currently, if the proxy receives two backends with the same metadata, one of the backends will get dropped because the backend metadata is used as the key in a hash map. Attempting to then randomly distribute requests to the backends can panic when selecting the now non-existent backend.

This is fixed by no longer using backend metadata as a hash map key, instead generating separate IDs that are stored in a vec to retain the declared order of backends while also being used to look up the backend and associated weight independetly.

Validated with new unit tests excercising duplicate backend keys, as well as a few around the invariants use to store the backends.

### CODE REVIEW

File: linkerd/distribute/src/params.rs [hunk 1]
```diff
@@ -16,17 +20,38 @@ pub enum Distribution<K> {
...
+
+impl KeyId {
+    pub(crate) fn new(idx: usize) -> Self {
+        Self { idx }
+    }
+}
+
+#[derive(Debug)]
+pub struct UnweightedKeys<K> {
+    ids: Vec<KeyId>,
```
Comment (sfleen) [inline linkerd/distribute/src/params.rs]: This is the key part of this change. It decouples the backend keys from the iteration order of the backends with a separate key type. I considered using something like [SlotMap](https://docs.rs/slotmap/latest/slotmap/), but since we don't modify or remove anything from these maps after they're created we don't need the guarantee of stable keys that it provides and the slight performance cost that comes with it.

File: linkerd/distribute/src/keys.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+impl KeyId {
+    pub(crate) fn new(idx: usize) -> Self {
+        Self { idx }
+    }
+}
+
+#[derive(Debug)]
+pub struct UnweightedKeys<K> {
+    ids: Vec<KeyId>,
```
Comment (sfleen) [inline linkerd/distribute/src/keys.rs:23]: This is the key part of this change. It decouples the backend keys from the iteration order of the backends with a separate key type. I considered using something like [SlotMap](https://docs.rs/slotmap/latest/slotmap/), but since we don't modify or remove anything from these maps after they're created we don't need the guarantee of stable keys that it provides and the slight performance cost that comes with it.

File: linkerd/distribute/src/keys.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        // self.keys have the same length
+        for id in &self.ids {
+            self.keys.get(id).hash(state);
+        }
+    }
+}
+
+impl<K> UnweightedKeys<K> {
+    pub(crate) fn new(iter: impl Iterator<Item = K>) -> Self {
+        let mut ids = Vec::new();
```
Comment (sfleen) [inline linkerd/distribute/src/keys.rs:76]: This is how the invariant of the IDs and keys being a 1:1 relation is guaranteed by only allowing construction via enumeration of the iterator over keys.

File: linkerd/distribute/src/params.rs [hunk 4]
```diff
@@ -64,46 +67,46 @@ impl<K> Default for Distribution<K> {
...
-            keys,
-            weights,
-        })))
+
+        weighted_keys.validate_weights()?;
+        Ok(Self::RandomAvailable(Arc::new(weighted_keys)))
     }
 
-    pub(crate) fn keys(&self) -> &[K] {
+    pub(crate) fn make_svc_from_keys<T>(
```
Comment (sfleen) [inline linkerd/distribute/src/params.rs:96]: This was another key change. Instead of exposing a slice of keys to create the backend services, this inverts the control to something similar to a combinator to decouple how they keys are iterated over and mapped into services.

File: linkerd/distribute/src/stack.rs [hunk 5]
```diff
@@ -55,16 +55,11 @@ where
...
         // Build the backends needed for this distribution, in the required
         // order (so that weighted indices align).
         let newk = self.inner.new_service(target);
-        let backends = dist
-            .keys()
-            .iter()
-            .map(|k| (k.clone(), newk.new_service(k.clone())))
-            .collect();
-        Distribute::new(backends, dist)
+        Distribute::new(dist, |k| newk.new_service(k.clone()))
```
Comment (sfleen) [inline linkerd/distribute/src/stack.rs:63]: And now this is much simpler and is decoupled from how the service keys are stored.

**Review by cratelyn — COMMENTED** [review]
here's a group of comments, _most_ of which are nits, and one [possible bug](https://github.com/linkerd/linkerd2-proxy/pull/3319#discussion_r1828506021)?

zooming out, i'm curious to tug at whether we can dodge the need to maintain a separate "keyspace" in order to account for duplicate keys.

i'll offer some more thoughts on that in a follow-up comment, but wanted to offer these comments in the meantime!

File: linkerd/distribute/src/service.rs [hunk 6]
```diff
@@ -88,25 +92,22 @@ where
...
                     if svc.poll_ready(cx)?.is_ready() {
-                        self.ready_idx = Some(idx);
+                        self.ready_idx = Some(id);
                         return Poll::Ready(Ok(()));
                     }
 
                     // Zero out the weight of the backend we just tried so that
                     // it's not selected again.
-                    match index.update_weights(&[(idx, &0)]) {
+                    match selector.disable_backend(id) {
```
[the remark this answers is not in the document]
Comment (cratelyn) [reply linkerd/distribute/src/service.rs:110]: figured this out, it's by design that this only disables the backend for the duration of this `loop {}`.

i'm going to mark this as resolved!

File: linkerd/distribute/src/service.rs [hunk 7]
```diff
@@ -196,11 +197,31 @@ mod tests {
...
+        assert!(svcs.is_empty());
+        dist
+    }
+
     #[test]
     fn empty_pending() {
         let mut dist_svc = mock::Spawn::new(Distribute::<&'static str, mock::Mock<(), ()>>::new(
             Default::default(),
-            Default::default(),
+            |_| unimplemented!(),
```
[the remark this answers is not in the document]
Comment (cratelyn) [reply linkerd/distribute/src/service.rs:224]: having spent some more time reading these tests, i like this change actually. it's a nice extra assurance that the test will panic if this ends up being called. that seems like a nice extra layer of assurance for a test about empty distributions never becoming available.

File: linkerd/distribute/src/keys.rs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    }
+}
+
+#[derive(Debug)]
+pub struct UnweightedKeys<K> {
+    ids: Vec<KeyId>,
+    keys: HashMap<KeyId, K>,
+}
+
+pub type WeightedKeys<K> = UnweightedKeys<WeightedKey<K>>;
```
[the remark this answers is not in the document]
Comment (cratelyn) [reply linkerd/distribute/src/keys.rs:27]: i've wrapped my head around this more this morning. i'm going to mark this as resolved!

File: linkerd/distribute/src/params.rs (hunk 4 above)
Comment (sfleen) [reply linkerd/distribute/src/params.rs]: Done.

File: linkerd/distribute/src/stack.rs [hunk 9]
```diff
@@ -55,16 +55,11 @@ where
     /// Referencing other keys causes a panic.
     fn new_service(&self, target: T) -> Self::Service {
         let dist = self.extract.extract_param(&target);
-        tracing::debug!(backends = ?dist.keys(), "New distribution");
+        tracing::debug!(backends = ?dist, "New distribution");
 
         // Build the backends needed for this distribution, in the required
         // order (so that weighted indices align).
```
[the remark this answers is not in the document]
Comment (sfleen) [reply linkerd/distribute/src/stack.rs]: Done

File: linkerd/distribute/src/params.rs [hunk 10]
```diff
@@ -64,46 +63,46 @@ impl<K> Default for Distribution<K> {
...
+        &self,
+        mut make_svc: impl FnMut(&K) -> T,
```
[the remark this answers is not in the document]
Comment (sfleen) [reply linkerd/distribute/src/params.rs]: It took a bit of fiddling to get the tests to agree but I like this solution better. Done.

**Review by cratelyn — APPROVED** [review]

**Comment by sfleen** [comment]
I've added an extra refactoring that creates separate structs for each distribution variant, which I think makes it a lot clearer and easier to understand.

File: linkerd/distribute/src/keys.rs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    distributions::{WeightedError, WeightedIndex},
+    prelude::Distribution as _,
+    Rng,
+};
+use std::{collections::HashMap, hash::Hash};
+
+/// Uniquely identifies a key/backend pair for a distribution. This allows
+/// backends to have the same key and still participate in request distribution.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct KeyId {
```
Comment (olix0r) [inline linkerd/distribute/src/keys.rs:11]: It looks like this type scope can be limited--we can keep this an internal implementation detail:
```suggestion
pub(crate) struct KeyId {
```


With this change, we would typically move the declaration down below the real pub types.

File: linkerd/distribute/src/keys.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    idx: usize,
+}
+
+impl KeyId {
+    pub(crate) fn new(idx: usize) -> Self {
+        Self { idx }
+    }
+}
```
Comment (olix0r) [inline linkerd/distribute/src/keys.rs:19]: nit: We tend to avoid mixing impls and type defs -- should generally move down below the other struct definitions.

File: linkerd/distribute/src/service/empty.rs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        tracing::debug!("empty distribution will never become ready");
+        Poll::Pending
+    }
+
+    #[inline]
+    fn call(&mut self, _: Req) -> Self::Future {
+        unreachable!("EmptySelection is never ready")
+    }
+}
```
Comment (olix0r) [inline linkerd/distribute/src/service/empty.rs:38]: This type doesn't really carry any weight -- we can easily collapse this case into the parent service

File: linkerd/distribute/src/service/empty.rs [hunk 14]
```diff
@@ -0,0 +1,38 @@
+use linkerd_stack::Service;
+use std::{
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+
+#[derive(Debug, Copy, Clone)]
+pub struct EmptySelection<S> {
```
Comment (olix0r) [inline linkerd/distribute/src/service/empty.rs:8]:
```suggestion
pub(crate) struct EmptySelection<S> {
```

File: linkerd/distribute/src/service/first.rs [hunk 15]
```diff
@@ -0,0 +1,80 @@
+use crate::keys::ServiceKeys;
+use linkerd_stack::{NewService, Service};
+use std::task::{Context, Poll};
+
+#[derive(Debug)]
+pub struct FirstAvailableSelection<S> {
```
Comment (olix0r) [inline linkerd/distribute/src/service/first.rs:6]: Not actually public:

```suggestion
pub(crate) struct FirstAvailableSelection<S> {
```

File: linkerd/distribute/src/service/random.rs [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+use rand::{distributions::WeightedError, rngs::SmallRng, SeedableRng};
+use std::{
+    collections::HashMap,
+    hash::Hash,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+pub struct RandomAvailableSelection<K, S> {
```
Comment (olix0r) [inline linkerd/distribute/src/service/random.rs:12]:
```suggestion
pub(crate) struct RandomAvailableSelection<K, S> {
```

File: linkerd/distribute/src/keys.rs [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+#[derive(Debug)]
+pub struct ServiceKeys<K> {
+    ids: Vec<KeyId>,
+    keys: HashMap<KeyId, K>,
```
Comment (olix0r) [inline linkerd/distribute/src/keys.rs:24]: Note that this crate already uses an AHashMap (faster hash algo), so it would be suitable to use that here as well
```suggestion
    keys: ahash::AHashMap<KeyId, K>,
```

File: linkerd/distribute/src/service/random.rs [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    keys: Arc<WeightedServiceKeys<K>>,
+    backends: HashMap<KeyId, S>,
```
Comment (olix0r) [inline linkerd/distribute/src/service/random.rs:14]: And here as well:
```suggestion
    backends: ashash::AHashMap<KeyId, S>,
```

**Review by olix0r — APPROVED** [review]
Overall approach looks good!

* We should `pub(crate)` structs that are not part of the public interface.
* We should use AHashMap to be consistent with the backend cache in this crate.
* tioli: The EmptySelection is trivial and can be eliminated.

File: linkerd/distribute/src/keys.rs (hunk 11 above)
Comment (sfleen) [reply linkerd/distribute/src/keys.rs]: Done

File: linkerd/distribute/src/keys.rs (hunk 12 above)
Comment (sfleen) [reply linkerd/distribute/src/keys.rs]: Done

File: linkerd/distribute/src/keys.rs (hunk 17 above)
Comment (sfleen) [reply linkerd/distribute/src/keys.rs:19]: Done

File: linkerd/distribute/src/service/empty.rs (hunk 14 above)
Comment (sfleen) [reply linkerd/distribute/src/service/empty.rs]: Done

File: linkerd/distribute/src/service/empty.rs (hunk 13 above)
Comment (sfleen) [reply linkerd/distribute/src/service/empty.rs]: Done

File: linkerd/distribute/src/service/first.rs (hunk 15 above)
Comment (sfleen) [reply linkerd/distribute/src/service/first.rs]: Done

File: linkerd/distribute/src/service/random.rs (hunk 16 above)
Comment (sfleen) [reply linkerd/distribute/src/service/random.rs]: Done

File: linkerd/distribute/src/service/random.rs (hunk 18 above)
Comment (sfleen) [reply linkerd/distribute/src/service/random.rs:14]: Done
